cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,451 @@
+      throw new IOException("Can't create Secure Store. ", e);
+    }
+    return ks;
+  }
+
+  /**
+   * Try loading from the user specified path, if that fails for any reason
+   * other than bad password then try loading from the backup path.
+   *
+   * @param path       Path to load from
+   * @param backupPath Backup path (_NEW)
+   * @throws IOException
+   */
+  private static void tryLoadFromPath(KeyStore keyStore, Path path, Path backupPath, char[] password)
+    throws IOException {
+    try {
+      loadFromPath(keyStore, path, password);
+      // Successfully loaded the keystore. No need to keep the backup file.
+      Files.deleteIfExists(backupPath);
+      LOG.info("Secure store loaded successfully.");
+    } catch (IOException ioe) {
+      // Try the backup path if the loading failed for any reason other than incorrect password.
+      if (!isBadOrWrongPassword(ioe)) {
+        // Mark the current file as CORRUPTED
+        Files.move(path, Paths.get(path.toString() + "_CORRUPTED_" + System.currentTimeMillis()));
+        // Try loading from the backup path
+        loadFromPath(keyStore, backupPath, password);
+        Files.move(backupPath, path);
+        LOG.warn("Secure store loaded successfully from " + backupPath + " since " + path + " was corrupted.");
+      } else {
+        // Failed due to bad password.
+        throw ioe;
+      }
+    }
+  }
+
+  /**
+   * Persist the keystore on the file system.
+   *
+   * During the flush the steps are
+   *  1. Mark existing _NEW as orphaned, it will exist only if something had failed in the last run.
+   *  2. Try to write the current keystore in a _NEW file.
+   *  3. If something fails then revert the key store to the old state and throw IOException.
+   *  4. If everything is OK then rename the _NEW to the main file.
+   *
+   * @throws IOException
+   */
+  private void flush() throws IOException {
+    Path newPath = constructNewPath(path);
+    writeLock.lock();
+    try {
+      // Might exist if a backup has been restored etc.
+      if (Files.exists(newPath)) {
+        Files.move(newPath, Paths.get(newPath.toString() + "_ORPHANED_" + System.currentTimeMillis()));
+      }
+
+      // Flush the keystore, write the _NEW file first
+      writeToNew(newPath);
+      // Do Atomic rename _NEW to CURRENT
+      Files.move(newPath, path, ATOMIC_MOVE);
+    } catch (IOException ioe) {
+      LOG.error("Failed to persist the key store.", ioe);
+      throw ioe;
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  private void writeToNew(Path newPath) throws IOException {
+    try (OutputStream fos = new DataOutputStream(Files.newOutputStream(newPath))) {
+      keyStore.store(fos, password);
+    } catch (KeyStoreException e) {
+      throw new IOException("Can't store keystore " + this, e);
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException(
+        "No such algorithm storing keystore " + this, e);
```

Comment:
nit: checkstyle's char limit for us is 120 chars. Is that set correctly in your environment? This could easily fit in one line. Sam comment below.